libs/unity/library/Runtime/Scripts/NativeRender/NativeVideoRenderer.cs

```diff
@@ -0,0 +1,219 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using UnityEngine;
+using Microsoft.MixedReality.WebRTC.UnityPlugin;
+using UnityEngine.UI;
+
+namespace Microsoft.MixedReality.WebRTC
+{
+	/// <summary>
+	/// This will render the video stream through native calls with DX11 or OpenGL, completely bypassing C# marshalling.
+	/// This provides a considerable performance improvement.
+	/// </summary>
+    [AddComponentMenu("MixedReality-WebRTC/Native Video Renderer")]
+    public class NativeVideoRenderer : MonoBehaviour
+    {
+	    [SerializeField]
+	    private RawImage _rawImage;
+
+        private RemoteVideoTrack _source;
+        private Material _videoMaterial;
+        private NativeRenderer _nativeRenderer;
+
+        private Texture2D _textureY;
+        private Texture2D _textureU;
+        private Texture2D _textureV;
+
+        private VideoFrameQueue<I420AVideoFrameStorage> _i420aFrameQueue;
+
+        private void Awake()
+        {
+	        NativeRenderingPluginUpdate.AddRef(this);
+        }
+
+        private void OnDestroy()
+        {
+	        NativeRenderingPluginUpdate.DecRef(this);
+
+	        // Teardown will also get called when the PeerConnection is destroyed.
+	        // So this is probably going to be called twice in many cases.
+	        TearDown();
+        }
+
+        private void Update()
+        {
+            if (_i420aFrameQueue != null)
+            {
+	            if (_i420aFrameQueue.TryDequeue(out I420AVideoFrameStorage frame))
+	            {
+		            StartNativeRendering((int)frame.Width, (int)frame.Height);
+	            }
+            }
+        }
+
+	    /// <summary>
+        /// Start rendering the passed source.
+        /// </summary>
+        /// <remarks>
+        /// Can be used to handle <see cref="VideoTrackSource.VideoStreamStarted"/> or <see cref="VideoReceiver.VideoStreamStarted"/>.
+        /// </remarks>
+        public void StartRendering(IVideoSource source)
+        {
+	        _source = source as RemoteVideoTrack;
+	        Debug.Assert(_source != null, "NativeVideoRender currently only supports RemoteVideoTack");
+
+	        switch (source.FrameEncoding)
+            {
+                case VideoEncoding.I420A:
+	                _i420aFrameQueue = new VideoFrameQueue<I420AVideoFrameStorage>(2);
+	                _source.I420AVideoFrameReady += I420AVideoFrameReady;
+                    break;
+                case VideoEncoding.Argb32:
+	                break;
+            }
+        }
+
+        private void I420AVideoFrameReady(I420AVideoFrame frame)
+        {
+	        // Clear the video display to not confuse the user who could otherwise
+	        // think that the video is still playing but is lagging/frozen.
+	        _i420aFrameQueue.Enqueue(frame);
```

Comment:
```suggestion
	        _i420aFrameQueue?.Enqueue(frame);
```
`_i420aFrameQueue = null` at line 89 might be called while this is still running, so this should avoid crashing on null. Alternatively the queue alive rather than be null-ed it (maybe it can be cleared just in case).

Reply:
added